relayer/chain/ethereum/writer.go

```diff
@@ -80,7 +115,35 @@ func (wr *Writer) writeLoop(ctx context.Context) error {
 					return fmt.Errorf("Invalid message")
 				}
 
-				err := wr.Write(ctx, &concreteMsg)
+				err := wr.WriteChannel(ctx, &concreteMsg)
+				if err != nil {
+					wr.log.WithError(err).Error("Error submitting message to ethereum")
+				}
+			}
+		}
+	}
+}
+
+func (wr *Writer) writeBeefyLoop(ctx context.Context) error {
+	ticker := time.NewTicker(5 * time.Second)
+	for {
+		select {
+		case <-ctx.Done():
+			return wr.onDone(ctx)
+		case <-ticker.C:
```

Comment:
still haven't got to the substrate reader code yet, but it looks like you're doing something like running a 5 second ticker to check for new CommitmentWitnessed events from the db, rather than reacting instantly to them as they come in?

can you explain the actual flow for the substrate reader -> db -> ethereum writer piece and your thinking behind it?

Reply:
The database is used as the intermediary for every step of the BEEFY justification relay lifecycle (see PR description) for consistency. The database's primary utility is storing the BEEFY justification's `ValidatorAddresses` and `SignedCommitement` so we can properly build the CompleteSignatureCommitment.

We could favor direct communication between the services and have the Relaychain listener forward the witnessed BEEFY justification directly to the Ethereum writer by introducing another channel to both services. To remove the 5-second cron job entirely, the Ethereum listener checks the database on each new block for any completable items, then forwards them directly to the Ethereum writer.

Reply:
That would still involve a database though, right?

Reply:
@philipstanislaus yes, a database is still required in order to store the BEEFY justification so we can build the CompleteSignatureCommitment in the final step. Is it possible to use the `currentId` from `InitialVerificationSuccessful` event to query the LightClientBridge's `ValidationData` and use that to build the CompleteSignatureCommitment instead?

For now, I refactored the Ethereum writer according to the approach described above in https://github.com/Snowfork/polkadot-ethereum/pull/314/commits/cb1c68c0d84516ae5c1ee6344cafc44794a4ea35 and removed the writer's 5-second cron job entirely by having the Ethereum listener fetch completable items and forward them to the writer as well.

Reply:
Included this as part of issue https://github.com/Snowfork/polkadot-ethereum/issues/340